### ISSUE TITLE
Requiring a file that uses a .jar is failing with Java StackOverflow in certain cases

### ISSUE THREAD

**Issue by fidothe** [jruby/jruby#5914]
### Environment

JRuby 9.2.8.0 (and a recent 9.2.9 snapshot)

macOS 10.14.6 (Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64)

JDKs 1.8, 12

When requiring a particular file (`saxon/item_type.rb`) from the Saxon-rb project (https://github.com/fidothe/saxon-rb), if the Saxon JAR is loaded using `jar-dependencies` generated `lib/saxon_jars.rb` file, everything is file. If the JAR is added to the class path and  then `import_package` is used without an intermediate `require`, a stack overflow is raised at the point `item_type.rb` attempts to use another class to generate values for a constant.

If that code is commented out (https://github.com/fidothe/saxon-rb/blob/46c0bc9f818d84e5a5f623418db87160179f6822/lib/saxon/item_type.rb#L118-L122), the error  doesn't happen.

There's a info on reproducing and a stack trace at https://gist.github.com/fidothe/5247bb8236e6dcd42d6bbc1d1b63ad2f.

**Comment by headius** [comment]
@mkristian You might have some cycles to look into this?

It appears that the trigger is the `const_missing` in the s9api.rb file. I'm pulling the bits now to try to reproduce and see what constant it can't find.

**Comment by headius** [comment]
Ok I don't have a solution, but I have a theory.

In [loader.rb](https://github.com/fidothe/saxon-rb/blob/master/lib/saxon/loader.rb#L96), you are using `include_package` to pull the `net.sf.saxon.s9api` package into the `Saxon::S9API`. In JRuby, because we can't query the JDK for a list of all classes in a given package,  `include_package` actually uses a `const_missing` hook to lazily load those classes.

However, in [s9api.rb](https://github.com/fidothe/saxon-rb/blob/master/lib/saxon/s9api.rb#L5) you define a new `const_missing`. This is very likely interfering with the `include_package` version.

I do not have an explanation for why it works with `jar-dependencies` but not directly on the classpath, but I suspect it's just causing code to load in a different order. In other words, when loading via `jar-dependencies`, the necessary classes are getting loaded and defined as constants before you replace the `include_package` version of `const_missing` with your own.

The following patch appears to work around the problem, but I would encourage you to find an alternative way to lazily load the Saxon library rather than having to chain `const_missing` like this.

```diff
diff --git a/lib/saxon/s9api.rb b/lib/saxon/s9api.rb
index 4dbc5e5..fef43ab 100644
--- a/lib/saxon/s9api.rb
+++ b/lib/saxon/s9api.rb
@@ -2,14 +2,11 @@ require 'saxon/loader'
 
 module Saxon
   module S9API
-    def self.const_missing(name)
-      Saxon::Loader.load!
-      begin
-        const_get(name)
-      rescue NameError
-        msg = "uninitialized constant Saxon::S9API::#{name}"
-        e = NameError.new(msg, name)
-        raise e
+    class << self
+      alias old_const_missing const_missing
+      def const_missing(name)
+        Saxon::Loader.load!
+        old_const_missing(name)
       end
     end
   end
```

@mkristian If you have any thoughts on why `jar-dependencies` would work better here, I'm all ears. 👂👂👂👂

**Comment by fidothe** [comment]
It was definitely beginning to smell like an ordering thing.

I think I know where this started. In the old just-XSLT version, `saxon-xslt`, I used `java_import` and explicitly imported every class I needed, not `include_package`. (https://github.com/fidothe/saxon-xslt/blob/52b4144f93302e1dc3e32864bd016421afa5a4fa/lib/saxon/loader.rb#L92-L98). I switched to `include_package` when I began `saxon-rb` because I was using so many more of the Java classes directly, and I clearly didn't remember that that would cause `const_missing` problems.

Lazily loading the library is hard - If you have a copy of Saxon EE, for example, you could have it on the classpath as a single .jar from Maven, directly downloaded from Saxonica, where it's split across two or three .jars. The naming scheme of the .jars is different as well - plain `saxon9[hpe]e.jar` in the direct download and maven style `Saxon-[HPE]E-9.X.X-X`. 

Being able to set the .jars that should be used at runtime seemed like a big win to me, and it certainly makes running an app that makes use of that easily runnable with `bundle exec` and `#!/usr/bin/env ruby`, rather than with `jruby -I blah -J-cp blah.jar script.rb`. I'm not sure I know how to do that without leaning on `const_missing`.

I will have a think and see what I can come up with. Thanks for taking a deeper look, and sorry for the hassle.

**Comment by fidothe** [comment]
I decided to stick with the basic approach, because I really can't think of a better starting point (would love to have a better idea), but now I'm keeping the `import_package` use in an anonymous module and pulling constants through into `Saxon::S9API` so I don't have to interfere with the `const_missing` defined by `import_package`. Thanks again!

**Comment by fidothe** [comment]
Reopened because I realised just now that the fact that my original code ever worked might itself be a bug, and I probably shouldn't be the one to make the closure call.

**Comment by headius** [comment]
@fidothe Yeah thanks for leaving it open. I'd like to know the answer to why it differs between the two modes before we call it resolved, and if something has regressed we'll want to figure out why.

**Comment by fidothe** [comment]
The main difference seems to be that going via `jar-dependencies` results in an explicit call to `Kernel.require` and the other way does not, relying on adding the external `.jar`s to the load path and letting `import_package` and `java_import` take the strain.

(This was because some saxon versions bundle ICU which might well already be on the classpath and I didn't know what requiring an already-loaded `.jar` would do)

**Comment by headius** [comment]
Fix or punt for 9.2.10.

**Comment by headius** [comment]
Ok, after more inspection of the logic involved, I'm still unsure why this would work in one mode and not another. I'm going to close this as invalid and maybe @mkristian has some clue about it.

In any case I don't believe there's anything to be fixed in JRuby.
